### PR TITLE
Add Cf compliant time to core_atmosphere

### PR DESCRIPTION
**Pull request by dimomatt**
CF compliance in time axis requires a unit of '<time unit> since YYYY-MM-DD hh:mm:ss'. This commit creates the 'Time' variable in the core_atmosphere Registry.xml, initializes it to 0, and changes the 'units' attribute to 'seconds since <start_timestamp>'. The start timestamp is modified from the initial xtime format to remove the underscore and replace it with a space to match the CF compliant time string format.

We can then start to update the Time variable at the same time as the xtime variable by adding 'dt' to the Time at time level 1, and saving it to time level 2. This produces a CF-compliant time variable that tracks xtime at each step.

### CODE REVIEW

File: src/core_atmosphere/mpas_atm_core.F [hunk 1]
```diff
@@ -262,6 +266,15 @@ function atm_core_init(domain, startTimeStamp) result(ierr)
          call mpas_pool_get_array(state, 'xtime', xtime, 1)
          xtime = startTimeStamp
 
+         ! Set the units to be cf compliant 'seconds since <cf-timestamp>'
+         call mpas_pool_get_field(state, 'Time', Time_field) 
+         call mpas_modify_att(Time_field % attLists(1) % attList, 'units', &
+              'seconds since ' // mpas_string_replace(startTimeStamp, '_', ' '))
```
Comment (mgduda) [inline src/core_atmosphere/mpas_atm_core.F:272]: So that we use a consistent reference time for restart runs, I think we'll need to use the `initial_time` variable rather than the start time. For example, if I do a cold-start run and simulate for two days, I get
```
	float Time(Time) ;
		Time:units = "seconds since 2010-10-23 00:00:00" ;
		Time:long_name = "CF convention compliant model valid time" ;
```
with values
```
 Time = 0, 21600, 43200, 64800, 86400, 108000, 129600, 151200, 172800 ;
```
However, if I restart from T+24h, I get
```
	float Time(Time) ;
		Time:units = "seconds since 2010-10-24 00:00:00" ;
		Time:long_name = "CF convention compliant model valid time" ;
```
with times
```
 Time = 21600, 43200, 64800, 86400 ;
```
in the model output. (Note that for a restart, the first history file record isn't written, hence no value of 0 for `Time`).

File: src/core_atmosphere/mpas_atm_core.F [hunk 2]
```diff
@@ -262,6 +266,15 @@ function atm_core_init(domain, startTimeStamp) result(ierr)
...
+
+         ! Initialize Time to 0
+         call mpas_pool_get_array(state, 'Time', Time, 1)
+         Time = 0.
```
Comment (mgduda) [inline src/core_atmosphere/mpas_atm_core.F:276]: Since `Time` is read from the restart file, I don't think we want to always reset it to zero here.

Comment (mgduda) [reply src/core_atmosphere/mpas_atm_core.F:276]: Following up on this, all real-valued variables declared in the `Registry.xml` file are set to zero after being allocated. But, I think we may want to consider changes to the `init_atmosphere` core (to be included as additional commits in this PR branch) that add a CF-compliant `Time` variable to the ICs, LBCs, and surface update files.

File: src/core_init_atmosphere/Registry.xml [hunk 3]
```diff
@@ -599,9 +601,10 @@
                         filename_interval="output_interval"
                         packages="lbcs"
                         immutable="true">
-
+
```
Comment (mgduda) [inline src/core_init_atmosphere/Registry.xml:604]: Please remove the added space here.

File: src/core_init_atmosphere/Registry.xml [hunk 4]
```diff
@@ -599,9 +601,10 @@
                         filename_interval="output_interval"
                         packages="lbcs"
                         immutable="true">
-
+ 
 			<var name="xtime"/>
-			<var_struct name="lbc_state"/>
+			<var name="Time"/>
+            <var_struct name="lbc_state"/>
```
Comment (mgduda) [inline src/core_init_atmosphere/Registry.xml:607]: I think the indentation here also needs fixing.

File: src/core_init_atmosphere/Registry.xml [hunk 5]
```diff
@@ -940,6 +943,9 @@
                 <var name="xtime" type="text" dimensions="Time" units="YYYY-MM-DD_hh:mm:ss"
                      description="Model valid time"/>
 
+                <var name="Time" type="real" dimensions="Time" units="Seconds since YYYY-MM-DD hh:mm:ss"
+                     desciption="CF convention compliant model valid time"/>
```
Comment (mgduda) [inline src/core_init_atmosphere/Registry.xml:947]: `desciption` should be `description`.

**Review by mgduda — CHANGES_REQUESTED** [review]
In addition to the comments, I think there should be some additional code in the init_atmosphere core to set the description for the `Time` variable to use the actual time of the initial conditions. Right now, the ICs file that's produced has the following `units`:
```
	float Time(Time) ;
		Time:units = "Seconds since YYYY-MM-DD hh:mm:ss" ;
```

**Comment by mgduda** [comment]
After more testing, I think there are a few additions that we may like to include in this PR:

- [ ] Add the CF-compliant `Time` variable to all default output streams in the `atmosphere` core
- [ ] Add the `standard_name` attribute with a value of `time` to the `Time` variable
- [ ] Ensure that the value of the `Time` variable is correct for each of the time slices in surface update files produced by the `init_atmosphere` core
- [ ] Ensure that the value of the `Time` variable is correct for each of the time slices in lateral boundary files produced by the `init_atmosphere` core

File: src/core_init_atmosphere/mpas_init_atm_surface.F [hunk 6]
```diff
@@ -44,29 +44,38 @@ subroutine init_atm_case_sfc(domain, dminfo, stream_manager, mesh, fg, state, di
...
+ start_time = mpas_get_clock_time(domain % clock, MPAS_START_TIME)
 
  do while (curr_time <= stop_time)
     call mpas_get_time(curr_time, dateTimeString=timeString)
     xtime = timeString
 
+    time_since_start = curr_time - start_time
+    call mpas_get_timeInterval(time_since_start, dt=dt)
+    Time = dt
+!
```
Comment (mgduda) [inline src/core_init_atmosphere/mpas_init_atm_surface.F:78]: I think it would be cleaner to remove the `!` that was added on this line -- just a blank line will suffice.

File: src/core_atmosphere/Registry.xml [hunk 7]
```diff
@@ -1478,6 +1482,9 @@
                 <var name="xtime" type="text" dimensions="Time" units="YYYY-MM-DD_hh:mm:ss"
                      description="Model valid time"/>
 
+                <var name="Time" type="real" dimensions="Time" units="Seconds since YYYY-MM-DD hh:mm:ss"
```
Comment (mgduda) [inline src/core_atmosphere/Registry.xml:1485]: Since the attribute that's ultimately written to files uses `seconds since ...`, would it be more consistent to un-capitalize `Seconds` here?

File: src/core_atmosphere/Registry.xml [hunk 8]
```diff
@@ -1478,6 +1482,9 @@
...
+                    description="CF convention compliant model valid time" standard_name="time"/>
```
Comment (mgduda) [inline src/core_atmosphere/Registry.xml:1486]: We could shorten this to something like "CF-compliant valid time".

File: src/core_init_atmosphere/Registry.xml [hunk 9]
```diff
@@ -940,6 +943,9 @@
                 <var name="xtime" type="text" dimensions="Time" units="YYYY-MM-DD_hh:mm:ss"
                      description="Model valid time"/>
 
+                <var name="Time" type="real" dimensions="Time" units="Seconds since YYYY-MM-DD hh:mm:ss"
+                     description="CF convention compliant model valid time" standard_name="time"/>
```
Comment (mgduda) [inline src/core_init_atmosphere/Registry.xml:947]: Similar comment as in `core_atmosphere/Registry.xml` regarding the capitalization of `Seconds` and the description.

File: src/core_atmosphere/dynamics/mpas_atm_time_integration.F [hunk 10]
```diff
@@ -299,6 +301,14 @@ subroutine atm_timestep(domain, dt, nowTime, itimestep, exchange_halo_group)
       call mpas_pool_get_array(state, 'xtime', xtime, 2)
       xtime = xtime_new
 
+      ! Get CF-compliant time at current timestep
+      call mpas_pool_get_array(state, 'Time', Time, 1)
+      Time_new = Time + dt
+
+      ! Write CF-compliant time for advanced timestep
+      call mpas_pool_get_array(state, 'Time', Time, 2)
+      Time = Time_new
```
Comment (mgduda) [inline src/core_atmosphere/dynamics/mpas_atm_time_integration.F:310]: An alternative to consider for this new code might be the following:
```
      ! Update CF-compliant time at end of timestep
      call mpas_pool_get_array(state, 'Time', Time, 1)
      call mpas_pool_get_array(state, 'Time', Time_new, 2)
      Time_new = Time + dt
```
There's one less line of code, and it seems clearer to me. But I recognize that this is purely subjective, so I wouldn't object to the code as it is in the PR. Just a different way to consider. (Note that the variant code here requires `Time_new` to be declared with the `pointer` attribute.)

**Review by mgduda — CHANGES_REQUESTED** [review]
Other than a few minor comments, I think this PR looks good from a correctness and functionality perspective.

File: src/core_atmosphere/dynamics/mpas_atm_time_integration.F (hunk 10 above)
Comment (dimomatt) [reply src/core_atmosphere/dynamics/mpas_atm_time_integration.F:310]: I think I'm going to leave it as is for now, but this is certainly a more concise way to do this. I tried to follow the style of the xtime update in the lines above, which maybe helps with clarity if you're transitioning from using the xtime variable to the Time variable?

Comment (mgduda) [reply src/core_atmosphere/dynamics/mpas_atm_time_integration.F:310]: In the case of the `xtime` update, the code is necessarily more complicated because we're updating a character string from a `real` and an `MPAS_Time_type`; on the other hand, in the case of `Time`, we're simply adding two `real` variables together. In my opinion, it's unnecessarily confusing to have the `Time` variable point to two different quantities only a few lines apart and to assign a value to `Time_new` only to then assign `Time_new` to `Time` (but not the same `Time` as a few lines earlier).

**Review by mgduda — APPROVED** [review]
